README.md
# PDF-PhotoBook
Perl processor for PDF


PDF-PhotoBook/Changes
Revision history for Perl extension PDF::PhotoBook.

0.01  Fri Nov  6 10:45:46 2009
	- original version; created by h2xs 1.23 with options
		-AX PDF::PhotoBook



test_suite/test.pl
#!/usr/bin/perl 
#===============================================================================
#
#         FILE:  test.pl
#
#        USAGE:  ./test.pl  
#
#  DESCRIPTION:  Test Suite File
#
#      OPTIONS:  ---
# REQUIREMENTS:  ---
#         BUGS:  ---
#        NOTES:  ---
#       AUTHOR:  Kail (), sickboy@anarchy.name
#      COMPANY:  Ergon
#      VERSION:  1.0
#      CREATED:  11/06/2009 11:16:42 AM
#     REVISION:  ---
#===============================================================================

use strict;
use warnings;

use PDF::PhotoBook;

my $book = new PDF::PhotoBook;

$book->SetTempDir('tmp');
    
my $layouts = [
            {'width'       => 500,
            'height'       => 800,
            'placeholders' => [
                            {
                            'type'        => 'image',          # Placeholder type image/text
                            'filepath'    => 'test.jpg',       # Image path
                            'width'       => 100,              # Placeholder width
                            'height'      => 300,              # Placeholder height
                            'x'           => 100,              # X cordinates
                            'y'           => 100,              # Y cordinates
                            'mask'        => 'test_mask.png',  # Optional. If we have frame we must apply transparency mask before adding it
                            'frame'       => 'test_frame.png', # Optional. If we have frame. 
                            'imageWidth'  => 100,              # For resizing the image
                            'imageHeight' => 300,              # For resizing the image
                            'imageX'      => 10,               # Optional. This is used to crop the image if we want to
                            'imageY'      => 20,               # Optional. This is used to crop the image if we want to
                            'flop'        => 1,                # Optional. If we want to flop the image
                            'degrees'     => 90,               # Optional. If we need a rotation
                            },
                            {
                            'type'     => 'text',           # Placeholder type image/text
                            'width'    => 300,              # Placeholder width
                            'height'   => 300,              # Placeholder height
                            'x'        => 0,                # X cordinates
                            'y'        => 500,              # Y cordinates
                            'ttf'      => 'arial.ttf',      # TTF font path
                            'text'     => 'Test text',      # Text content
                            'color'    => '#000000',        # Font color
                            'size'     => '32',             # Font size
                            'align'    => 'center',         # Text align in the placeholder - left/right/center
                            },
                        ],
            } 
            ]; 
$book->AddLayouts($layouts);
$book->Create('test.pdf');


base/Image.pm
package PhotoAlbums::PhotoAlbums::Utils::Image;
use Image::Magick;
use PDF::Reuse;
use strict;

sub new {
	my ($class, $args) = @_;
	my $self = {};

	bless($self, $class);
	
	$self->{'registry'} = PhotoAlbums::PFrame1_2::Registry->instance();
	$self->{'config'} = $self->{'registry'}->{'store'}->{'config'} if !$self->{'config'};
	return $self;
}

sub createMain{
	my ($self, $args) = @_;
	#my ($self, $template) = @_;

	#print STDERR "\033[34mCreate main " . $template . "\033[0m\n";
	print STDERR "\033[34mCreate main \033[0m\n";
	$self->{'template'} = Image::Magick->new();
	#$self->{'template'}->ReadImage( "png:" . $template);
	# Set correct size
	$self->{'template'}->Set('size' => $self->{'width'} . 'x' . $self->{'height'});
	$self->{'template'}->ReadImage( "xc: white");
	return 1;
}

sub setSizes{
	my ($self, $width, $height) = @_;
	
	$self->{'width'}  = $width;
	$self->{'height'} = $height;

	return 1;
}

sub checkImageFormat{
	my ($self, $args) = @_;
	
	$self->{'userImage'} = Image::Magick->new();
	my ($width, $height, $size, $format) = $self->{'userImage'}->Ping($args->{'userImage'});
	
	if (!$format){
		return 0;
	}
	$self->{'userImageType'} = $format;
	return 1;
}

sub applyMask{
	my ($self, $args) = @_;

	# Done above
	#$self->{'userImage'} = Image::Magick->new() if !$self->{'userImage'};

	print STDERR "IMAGE FORMAT: " . $self->{'userImageType'} . "\n";
	print STDERR $self->{'userImage'}->ReadImage( $self->{'userImageType'} . ":" . $args->{'userImage'});
	$self->{'userImage'}->Set( alpha => 'set');
	
	# Rotate image
	if ($args->{'degrees'} && $args->{'degrees'} > 0){
		print STDERR "Rotating to: " . $args->{'degrees'} . " degrees \n";
		$self->{'userImage'}->Rotate(degrees => $args->{'degrees'}, background => 'none');
	}

	# Zoom, Crop etc...
	$self->{'userImage'}->Resize(geometry => $args->{'imageWidth'} . 'x' . $args->{'imageHeight'});

	#$args->{'imageX'} = '+' . $args->{'imageX'} if $args->{'imageX'} >= 0;
	#$args->{'imageY'} = '+' . $args->{'imageY'} if $args->{'imageY'} >= 0;

	$args->{'imageX'} = $args->{'imageX'} * -1;
	$args->{'imageY'} = $args->{'imageY'} * -1;
	print STDERR $args->{'imageWidth'} . 'x' . $args->{'imageHeight'} . ' X = ' . $args->{'imageX'} . ' Y = ' . $args->{'imageY'} . "\n";
	#my $err = $self->{'framed'}->Crop($args->{'placeWidth'} . 'x' . $args->{'placeHeight'} . $args->{'imageX'}  . $args->{'imageY'});
	my $err = $self->{'userImage'}->Crop('width'  => $args->{'placeWidth'},
					  'height' => $args->{'placeHeight'},
					  'x'      => $args->{'imageX'},
					  'y'      => $args->{'imageY'},
				  );
	warn $err if $err;
	
	# Flop image
	if ($args->{'flop'}){
		$self->{'userImage'}->Flop();
	}
	if ($args->{'mask'}){
		$self->{'mask'} = Image::Magick->new();
		$self->{'mask'}->ReadImage( "png:" . $args->{'mask'});
		$self->{'mask'}->Resize(width => $args->{'placeWidth'}, height => $args->{'placeHeight'});
	
		$self->{'userImage'}->Composite( image => $self->{'mask'},
		                  compose => 'DstIn',
		                  gravity => 'Center',
		                  quality => 100,
		                );
		push (@{$self->{'garbage'}}, $self->{'mask'});
		#$self->{'userImage'}->Write("png:masked.png");
	} 
	push (@{$self->{'garbage'}}, $self->{'userImage'});
	return 1;
}

sub applyFrame{
	my ($self, $args) = @_;
	
	print STDERR "   Appling frame \n";
	if ($args->{'frame'}){
		$self->{'framed'} = Image::Magick->new();
		$self->{'framed'}->ReadImage( "png:" . $args->{'frame'});

		#$self->{'framed'}->Resize(geometry => $args->{'placeWidth'} . 'x' . $args->{'placeHeight'});
		$self->{'framed'}->Resize(width => $args->{'placeWidth'}, height => $args->{'placeHeight'});

		$self->{'userImage'}->Composite( image => $self->{'framed'},
		                  compose => 'over',
		                  gravity => 'Center',
		                  quality => 100,
		                );
		push (@{$self->{'garbage'}}, $self->{'framed'});
	} 
	#$self->{'framed'}->Write("png:framed.pdf");
	return 1;
}


sub paint{
	my ($self, $args) = @_;

	print STDERR "   Painting \n";

	# Parametrize this!!!
#	$self->{'userImage'}->Rotate(degrees => 0, background => 'none');
#
#	# Zoom, Crop etc...
#	$self->{'userImage'}->Resize(geometry => $args->{'imageWidth'} . 'x' . $args->{'imageHeight'});
#	
	#$args->{'imageX'} = '+' . $args->{'imageX'} if $args->{'imageX'} >= 0;
	#$args->{'imageY'} = '+' . $args->{'imageY'} if $args->{'imageY'} >= 0;

#	$args->{'imageX'} = $args->{'imageX'} * -1;
#	$args->{'imageY'} = $args->{'imageY'} * -1;
#	print STDERR $args->{'imageWidth'} . 'x' . $args->{'imageHeight'} . ' X = ' . $args->{'imageX'} . ' Y = ' . $args->{'imageY'} . "\n";
	#my $err = $self->{'userImage'}->Crop($args->{'placeWidth'} . 'x' . $args->{'placeHeight'} . $args->{'imageX'}  . $args->{'imageY'});
#	my $err = $self->{'userImage'}->Crop('width'  => $args->{'placeWidth'},
#					  'height' => $args->{'placeHeight'},
#					  'x'      => $args->{'imageX'},
#					  'y'      => $args->{'imageY'},
#				  );
#	warn $err if $err;
	# TODO Parametrize this
	#$self->{'userImage'}->Flop();
	print STDERR $self->{'template'}->Composite(
					image => $self->{'userImage'},
					compose => 'Over',
					#rotate  => 20,
					gravity => 'NorthWest',
					geometry=> '+' . $args->{'placeX'} . '+' . $args->{'placeY'},
					#geometry=>'10x10',
					#quality => 100,
					);
	return 1;
}

sub createLayoutPdf{
	my ($self, $layputPdf) = @_;

	$self->{'pdf'} = Image::Magick->new();
	# Set correct size
	$self->{'pdf'}->Set('size' => $self->{'width'} . 'x' . $self->{'height'});
	$self->{'pdf'}->ReadImage( "xc: white");

	$self->{'pdf'}->Composite( image => $self->{'template'},
	                 compose => 'Over',
       		         #mask    => $main,
       		         x => 0,
       		         y => 0,
               		 gravity => 'East',
               		 quality => 100,
        );

	$self->{'pdf'}->Write("pdf:" . $layputPdf);
	push (@{$self->{'garbage'}}, $self->{'pdf'});
	push (@{$self->{'pdfs'}}, $layputPdf);
	return 1;
}

sub composePdf{
	my ($self, $args) = @_;
	
	print STDERR "\033[34mComposing PDF " . $args->{'output'} . " \033[0m\n";
	prFile({'Name' => $args->{'output'},
	        'FitWindow'    => 1,
	        'CenterWindow' => 1,
	        'HideToolbar'  => 1,
	        'HideMenubar'  => 1,
	        'HideWindowUI' => 1,
	        'FitWindow'    => 1,
        });
	#prMbox ( 0,0,0,0 );
	# Set cprrect size
	prMbox(0, 0, $self->{'width'}, $self->{'height'});
	prCompress(1);

	foreach (@{$self->{'pdfs'}}){
		prImage({'file'  => $_,
			'adjust' => 1,
			});
		prPage();
	}
	prEnd;	
	foreach (@{$self->{'pdfs'}}){
		unlink $_ if -e $_;
	}
	return 1;
}

sub addText{
	my ($self, $args) = @_;

	print STDERR "   Add text \n";

	use Data::Dumper;
	warn Dumper $args;
	# Set correct font path, set fontsize

	if ($args->{'align'} eq 'Center'){
		$args->{'placeX'} += ($args->{'placeWidth'}/2);
	} elsif ($args->{'align'} eq 'Right'){
		$args->{'placeX'} += $args->{'placeWidth'};
	}

	print STDERR $self->{'template'}->Annotate(
				  font      => $args->{'ttf'},
				  #pen       => 'black',
                                  fill      => ($args->{'fontColor'} || '#000000'),
                                  pointsize => ($args->{'fontSize'}  || 0),
				  #stroke    => '#000000',
				  #strokewidth => 5,
				  #undercolor => '#000000',
                                  #gravity   => 'East',
				  #geometry  => '+' . $args->{'placeWidth'} . '+' . $args->{'placeHeight'},
                                  text      => ($args->{'text'} || ''),
				  align     =>  $args->{'align'},
				  #gravity   => 'East',
				  x => $args->{'placeX'},
				  y => $args->{'placeY'},
				);
#	print STDERR $self->{'template'}->Draw(
#				primitive=>'text',
#				  font      => $args->{'ttf'},
#                                  #pen       => 'black',
#                                 fill      => ($args->{'fontColor'} || '#000000'),
#				 'pointsize' => ($args->{'fontSize'}  || 0),
#				  stroke    => '#000000',
#				  'strokewidth' => 5,
#				  #decorate => 'UnderLine',
#				  #undercolor => '#000000',
#                                  #gravity   => 'East',
#				  #geometry  => '+' . $args->{'placeWidth'} . '+' . $args->{'placeHeight'},
#                                  text      => ($args->{'text'} || ''),
#				  #align     =>  $args->{'align'},
#				  #gravity   => 'East',
#				  x => 100 || $args->{'placeX'},
#				  y => 100 || $args->{'placeY'},
#				);
	#$self->{'template'}->Composite( 
		     #image       => $self->{'template'},
                     #compose    =>  'Atop',
                     #mask        => $self->{'template'},
       #              gravity     => 'East',
       #              quality     => 100,
                     #interpolate => 'bilinear',
       #              );
	return 1;
}

sub write{
	my ($self, $args) = @_;

	print STDERR "   Writing text \n";
	return 1;
}

sub cleanUp{
	my ($self, $args) = @_;

	foreach (@{$self->{'garbage'}}){
		undef $_;
	}
	return 1;
}

1;


base/OrderProcessor.pm
package PhotoAlbums::PhotoAlbums::Controller::OrderProcessor;
use PhotoAlbums::PhotoAlbums::Model::OrderAlbumLayouts;
use PhotoAlbums::PhotoAlbums::Model::Orders;
use PhotoAlbums::PhotoAlbums::Utils::Image;
use base "PhotoAlbums::PFrame1_2::Controller::Default";
use strict;

sub new {
	my ($class, $args) = @_;
	my $self = $class->SUPER::new();

	bless($self, $class);
	return $self;
}

sub init{
	my ($self, $args) = @_;
	my $orders        = undef;
   
	$self->{'orders'} = new PhotoAlbums::PhotoAlbums::Model::Orders;
	while (1){
		$orders = $self->{'orders'}->fetchAll({'filter' => {'IsProcessed' => 'N', 'Status' => 'Approved'}});
		foreach (@{$orders}){
			$self->_processOrder($_->[0], $_->[1], {
								'coverWidth'  => $_->[3], 
								'coverHeight' => $_->[4],
								'width'       => $_->[5],
								'height'      => $_->[6],
								'dpi'         => $_->[7],
								});
		}
		print STDERR "Sleeping...\n";
		sleep (3);
	}
	return 1;
}

sub _processOrder{
	my ($self, $orderId, $userId, $args) = @_;
	my $orders           = new PhotoAlbums::PhotoAlbums::Model::Orders;
	my $orderLayouts     = new PhotoAlbums::PhotoAlbums::Model::OrderAlbumLayouts;
	my $painter          = new PhotoAlbums::PhotoAlbums::Utils::Image;
	my $orderData        = undef;
	my $layoutPdf        = undef;
	my $orderPdf         = undef;
	my $layoutWidth      = undef;
	my $layoutHeight     = undef;
	
	$orderData = $self->_prepareOrderData($orderLayouts->fetchAllByOrder({'filter' => {'OrderID' => $orderId}}));

	foreach (sort {$orderData->{$a}->{'pageNum'} <=> $orderData->{$b}->{'pageNum'}} keys %{$orderData}){
		print STDERR "\n\033[31mProcessing layout " . $_ . "...\033[0m\n";
		#$painter->createMain($self->{'config'}->{'PNG_TEMPLATES_PATH'} . $orderData->{$_}->{'template'});
		if ($orderData->{$_}->{'pageNum'} == 0){
			# We have cover here
			$layoutWidth  = $self->_mm2pixel($args->{'dpi'}, $args->{'coverWidth'});
			$layoutHeight = $self->_mm2pixel($args->{'dpi'}, $args->{'coverHeight'});
		} else {
			$layoutWidth  = $self->_mm2pixel($args->{'dpi'}, $args->{'width'});
			$layoutHeight = $self->_mm2pixel($args->{'dpi'}, $args->{'height'});
		}
		$painter->setSizes($layoutWidth, $layoutHeight);
		$painter->createMain({'width' => $layoutWidth, 'height' => $layoutHeight});
		foreach (@{$orderData->{$_}->{'placeholders'}}){
			print STDERR "\nProcessing placeholder...\n";
			if ($_->{'IID'} && $_->{'IID'} > 0 && $_->{'ImageName'} && -e ($self->{'config'}->{'USER_IMAGES_PATH'} . $userId . '/' . $_->{'ImageName'})){
				if (!$painter->checkImageFormat({'userImage' => ($self->{'config'}->{'USER_IMAGES_PATH'} . $userId . '/' . $_->{'ImageName'} || '')})){
					print STDERR "Image in unknown format: " . $self->{'config'}->{'USER_IMAGES_PATH'} . $userId . '/' . $_->{'ImageName'} . "\n";
					next;
				}
				$painter->applyMask({
						'userImage' => ($self->{'config'}->{'USER_IMAGES_PATH'} . $userId . '/' . $_->{'ImageName'} || ''),
						'mask'      => ($_->{'Mask'} ? $self->{'config'}->{'FRAMES_IMAGES_PATH'} . $_->{'Mask'} : ''),
						'placeWidth'  => $self->_mm2pixel($args->{'dpi'}, $_->{'LW'}),
						'placeHeight' => $self->_mm2pixel($args->{'dpi'}, $_->{'LH'}),
						'placeX'      => $self->_mm2pixel($args->{'dpi'}, $_->{'LX'}),
						'placeY'      => $self->_mm2pixel($args->{'dpi'}, $_->{'LY'}),
						'imageWidth'  => $self->_mm2pixel($args->{'dpi'}, $_->{'IW'}),
						'imageHeight' => $self->_mm2pixel($args->{'dpi'}, $_->{'IH'}),
						'imageX'      => $self->_mm2pixel($args->{'dpi'}, $_->{'IX'}),
						'imageY'      => $self->_mm2pixel($args->{'dpi'}, $_->{'IY'}),
						'flop'        => ($_->{'LIFL'} eq 'Y' ? 1 : undef),
						'degrees'     => $_->{'IR'},
					     	    });
				$painter->applyFrame({'frame'  => ($_->{'FrameImage'} ? $self->{'config'}->{'FRAMES_IMAGES_PATH'} . $_->{'FrameImage'} : undef),
						'placeWidth'  => $self->_mm2pixel($args->{'dpi'}, $_->{'LW'}),
						'placeHeight' => $self->_mm2pixel($args->{'dpi'}, $_->{'LH'}),
					      	     });
				# Additional params here!!!
				$painter->paint({
						'placeX'      => $self->_mm2pixel($args->{'dpi'}, $_->{'LX'}),
						'placeY'      => $self->_mm2pixel($args->{'dpi'}, $_->{'LY'}),
						});
			} elsif ($_->{'LTID'} && $_->{'LTID'} > 0){
				if ($_->{'Content'} && $_->{'Content'} ne ''){
					if (!$_->{'TTF'}){
						print STDERR "Missing param TTF file\n";
						next;
					}
					if (!-e $self->{'config'}->{'FONTS_PATH'} . $_->{'TTF'}){
						print STDERR "Unknown TTF font: " . $self->{'config'}->{'FONTS_PATH'} . $_->{'TTF'} . "\n";
						next;
					}
					$painter->addText({
							'placeWidth'  => $self->_mm2pixel($args->{'dpi'}, $_->{'LW'}),
							'placeHeight' => $self->_mm2pixel($args->{'dpi'}, $_->{'LH'}),
							'placeX'      => $self->_mm2pixel($args->{'dpi'}, $_->{'LX'}),
							'placeY'      => $self->_mm2pixel($args->{'dpi'}, $_->{'LY'}),
							'ttf'         => $self->{'config'}->{'FONTS_PATH'} . $_->{'TTF'},
							'fontColor'   => $_->{'FontColor'},
							'align'       => $_->{'Align'},
							'text'        => $_->{'Content'},
							'fontSize'    => $_->{'FontSize'},
							});
					$painter->write();
				}
			}
		}
		$layoutPdf = $self->{'config'}->{'TEMP_DIR'} . sprintf($self->{'config'}->{'LAYOUT_PDF'}, $_) . '.' . $self->{'config'}->{'PDF_EXT'};
		$painter->createLayoutPdf($layoutPdf);
		$painter->cleanUp();
	}
	$orderPdf = sprintf($self->{'config'}->{'ALBUM_PDF'}, $orderId) . '.' . $self->{'config'}->{'PDF_EXT'};
	$painter->composePdf({'output' => $self->{'config'}->{'ORDER_ALBUMS_PATH'} . $orderPdf});
	$orders->update({'id' => $orderId, 'fields' => {'PdfAlbum' => $orderPdf, 
							'IsProcessed' => 'Y',
						}});
	return 1;
}

sub _mm2pixel{
	my ($self, $dpi, $mm) = @_;

	# 0.0393700787 mm per inch
	return int (($mm * 0.0393700787) * $dpi);
}

sub _prepareOrderData{
	my ($self, $data) = @_;
	my $orderData     = {};

	#foreach (sort {$a->{'PageNum'} <=> $b->{'PageNum'}} values %{$data}){
	foreach (sort { $a->{'PageNum'} <=> $b->{'PageNum'} || $a->{'ZIndex'} <=> $b->{'ZIndex'} } values %{$data}){
		#$orderData->{$_->{'LayoutID'}}->{'template'} = $_->{'Template'};
		$orderData->{$_->{'OrderAlbumLayoutID'}}->{'pageNum'} = $_->{'PageNum'} if !$orderData->{$_->{'OrderAlbumLayoutID'}}->{'pageNum'};
		push (@{$orderData->{$_->{'OrderAlbumLayoutID'}}->{'placeholders'}}, $_);
	}
	
	return $orderData;
}

1;


PDF-PhotoBook/lib/PDF/PhotoBook.pm
package PDF::PhotoBook;

use 5.010000;
use strict;
use warnings;
use Image::Magick;
use PDF::Reuse;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration   use PDF::PhotoBook ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
   
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
   
);

our $VERSION = '0.01';


# Preloaded methods go here.

sub new {
   my ($class, $args) = @_;
   my $self = {};

   bless($self, $class);
   
   return $self;
}

# Public methods
sub SetTempDir{
    my ($self, $dir) = @_;

    $self->{'tmpdir'} = $dir || '/tmp';
    return 1;
}

sub AddLayouts{
    my ($self, $layouts) = @_;

    $self->{'layouts'} = $layouts || die('Missing param layouts for AddLayouts');
    return 1;
}

sub Create{
    my ($self, $pdf) = @_;

    foreach (@{$self->{'layouts'}}){

        $self->SetSizes($_->{'width'}, $_->{'height'});
        $self->CreatePage;
        foreach (@{$_->{'placeholders'}}){
            if ($_->{'type'} eq 'image'){
                if (!$self->CheckImageFormat({'userImage' => $_->{'filepath'}})){
                    warn "Image in unknown format: " . $_->{'filepath'} . "\n";
                    next;
                }
                $self->ApplyMask({
                        'userImage'   => $_->{'filepath'},
                        'mask'        => $_->{'mask'},
                        'placeWidth'  => $_->{'width'},
                        'placeHeight' => $_->{'height'},
                        'placeX'      => $_->{'x'},
                        'placeY'      => $_->{'y'},
                        'imageWidth'  => $_->{'imageWidth'},
                        'imageHeight' => $_->{'imageHeight'},
                        'imageX'      => ($_->{'imageX'}   || 0),
                        'imageY'      => ($_->{'imageY'} || 0),
                        'flop'        => $_->{'flop'},
                        'degrees'     => $_->{'degrees'},
                        });
                $self->ApplyFrame({'frame'  => $_->{'frame'},
                        'placeWidth'  => $_->{'width'},
                        'placeHeight' => $_->{'height'},
                                      });
                # Additional params here!!!
                $self->Paint({
                        'placeX'      => $_->{'x'},
                        'placeY'      => $_->{'y'},
                        });
            } elsif ($_->{'type'} eq 'text'){
                if (!$_->{'text'} || $_->{'text'} eq ''){
                    warn "Missing param text\n";
                }
                if (!-e $_->{'ttf'}){
                    print STDERR "File not found TTF font: " . $_->{'ttf'} . "\n";
                    next;
                }
                $self->Write({
                            'placeWidth'  => $_->{'width'},
                            'placeHeight' => $_->{'height'},
                            'placeX'      => $_->{'x'},
                            'placeY'      => $_->{'y'},
                            'ttf'         => $_->{'ttf'},
                            'fontColor'   => $_->{'color'},
                            'align'       => $_->{'align'},
                            'text'        => $_->{'text'},
                            'fontSize'    => $_->{'size'},
                            });
            }
        }
        $self->{'tempdir'} ||= '/tmp';
        my $tmpPdf = $self->{'tempdir'} . time . ".pdf";
        $self->CreateLayoutPdf($tmpPdf);
        $self->CleanUp();
    }    
    $self->ComposePdf($pdf);
    return 1;
}

# Private methods
sub CreatePage{
   my ($self, $args) = @_;

   $self->{'template'} = Image::Magick->new();
   $self->{'template'}->Set('size' => $self->{'width'} . 'x' . $self->{'height'});
   $self->{'template'}->ReadImage( "xc: white");
   return 1;
}

sub SetSizes{
   my ($self, $width, $height) = @_;
   
   $self->{'width'}  = $width;
   $self->{'height'} = $height;
   return 1;
}

sub CheckImageFormat{
   my ($self, $args) = @_;
   
   $self->{'userImage'} = Image::Magick->new();
   my ($width, $height, $size, $format) = $self->{'userImage'}->Ping($args->{'userImage'});
   
   if (!$format){
      return 0;
   }
   $self->{'userImageType'} = $format;
   $self->{'imageWidth'}    = $width;
   $self->{'imageHeight'}   = $height;
   return 1;
}

sub ApplyMask{
   my ($self, $args) = @_;

   $self->{'userImage'}->ReadImage( $self->{'userImageType'} . ":" . $args->{'userImage'});
   $self->{'userImage'}->Set( alpha => 'set');
   
   # Zoom, Crop etc...
   # Rotate image
   if ($args->{'degrees'} && $args->{'degrees'} > 0){
      $self->{'userImage'}->Rotate(degrees => $args->{'degrees'}, background => 'none');
   }
   if ($args->{'imageWidth'} && $args->{'imageHeight'}){
       $self->{'userImage'}->Resize(geometry => $args->{'imageWidth'} . 'x' . $args->{'imageHeight'});
   }

   $args->{'imageX'} = $args->{'imageX'} * -1;
   $args->{'imageY'} = $args->{'imageY'} * -1;
   my $err = $self->{'userImage'}->Crop('width'  => $args->{'placeWidth'},
                 'height' => $args->{'placeHeight'},
                 'x'      => $args->{'imageX'},
                 'y'      => $args->{'imageY'},
              );
   warn $err if $err;
   
   # Flop image
   if ($args->{'flop'}){
      $self->{'userImage'}->Flop();
   }
   if ($args->{'mask'}){
      $self->{'mask'} = Image::Magick->new();
      $self->{'mask'}->ReadImage( "png:" . $args->{'mask'});
      $self->{'mask'}->Resize(width => $args->{'placeWidth'}, height => $args->{'placeHeight'});
   
      $self->{'userImage'}->Composite( image => $self->{'mask'},
                        compose => 'DstIn',
                        gravity => 'Center',
                        quality => 100,
                      );
      push (@{$self->{'garbage'}}, $self->{'mask'});
   } 
   push (@{$self->{'garbage'}}, $self->{'userImage'});
   return 1;
}

sub ApplyFrame{
   my ($self, $args) = @_;
   
   if ($args->{'frame'}){
      $self->{'framed'} = Image::Magick->new();
      $self->{'framed'}->ReadImage( "png:" . $args->{'frame'});

      $self->{'framed'}->Resize(width => $args->{'placeWidth'}, height => $args->{'placeHeight'});
      $self->{'userImage'}->Composite( image => $self->{'framed'},
                        compose => 'over',
                        gravity => 'Center',
                        quality => 100,
                      );
      push (@{$self->{'garbage'}}, $self->{'framed'});
   } 
   return 1;
}


sub Paint{
   my ($self, $args) = @_;

   print STDERR $self->{'template'}->Composite(
               image => $self->{'userImage'},
               compose => 'Over',
               #rotate  => 20,
               gravity => 'NorthWest',
               geometry=> '+' . $args->{'placeX'} . '+' . $args->{'placeY'},
               #geometry=>'10x10',
               );
   return 1;
}

sub CreateLayoutPdf{
   my ($self, $layputPdf) = @_;

   $self->{'pdf'} = Image::Magick->new();
   # Set correct size
   $self->{'pdf'}->Set('size' => $self->{'width'} . 'x' . $self->{'height'});
   $self->{'pdf'}->ReadImage( "xc: white");

   $self->{'pdf'}->Composite( image => $self->{'template'},
                    compose => 'Over',
                      x => 0,
                      y => 0,
                      gravity => 'East',
                      quality => 100,
        );

   $self->{'pdf'}->Write("pdf:" . $layputPdf);
   push (@{$self->{'garbage'}}, $self->{'pdf'});
   push (@{$self->{'pdfs'}}, $layputPdf);
   return 1;
}

sub ComposePdf{
   my ($self, $pdf) = @_;
   
   prFile({'Name' => $pdf,
           'FitWindow'    => 1,
           'CenterWindow' => 1,
           'HideToolbar'  => 1,
           'HideMenubar'  => 1,
           'HideWindowUI' => 1,
           'FitWindow'    => 1,
        });
   # Set size
   prMbox(0, 0, $self->{'width'}, $self->{'height'});
   prCompress(1);

   foreach (@{$self->{'pdfs'}}){
      prImage({'file'  => $_,
         'adjust' => 1,
         });
      prPage();
   }
   prEnd;   
   foreach (@{$self->{'pdfs'}}){
      unlink $_ if -e $_;
   }
   return 1;
}

sub Write{
   my ($self, $args) = @_;

   if ($args->{'align'} eq 'Center'){
      $args->{'placeX'} += ($args->{'placeWidth'}/2);
   } elsif ($args->{'align'} eq 'Right'){
      $args->{'placeX'} += $args->{'placeWidth'};
   }

   print STDERR $self->{'template'}->Annotate(
               font      => $args->{'ttf'},
               #pen       => 'black',
               fill      => ($args->{'fontColor'} || '#000000'),
               pointsize => ($args->{'fontSize'}  || 0),
               #stroke    => '#000000',
               #strokewidth => 5,
               #undercolor => '#000000',
               #gravity   => 'East',
               #geometry  => '+' . $args->{'placeWidth'} . '+' . $args->{'placeHeight'},
               text      => ($args->{'text'} || ''),
               align     =>  $args->{'align'},
               #gravity   => 'East',
               x         => $args->{'placeX'},
               y         => $args->{'placeY'},
               antialias => 1,
            );
   return 1;
}

sub CleanUp{
   my ($self, $args) = @_;

   foreach (@{$self->{'garbage'}}){
      undef $_;
   }
   return 1;
}

1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

PDF::PhotoBook - Perl extension for building amazing photobooks ready to be printed out on any printable mashine.

=head1 SYNOPSIS

  use PDF::PhotoBook;
  my $book = new PDF::PhotoBook;
  $book->SetTempDir('/tmp'); # Optional. Default is /tmp
  my $layouts = [
         {'width'       => 500,
         'hieght'       => 800,
         'placeholders' => [
                     {
                     'type'        => 'image',          # Placeholder type image/text
                     'filepath'    => 'test.jpg',       # Image path
                     'width'       => 100,              # Placeholder width
                     'height'      => 300,              # Placeholder height
                     'x'           => 100,              # X cordinates
                     'y'           => 100,              # Y cordinates
                     },
                     {
                     'type'     => 'text',      # Placeholder type image/text
                     'width'    => 100,         # Placeholder width
                     'height'   => 300,         # Placeholder height
                     'x'        => 0,           # X cordinates
                     'y'        => 500,         # Y cordinates
                     'text'     => 'Test text', # Text content
                     'ttf'      => 'arial.ttf', # TTF font path
                     'color'    => '#000000',   # Font color
                     'size'     => '22',        # Font size
                     },
                  ],
         }, 
         {'width'       => 500,
         'hieght'       => 800,
         'placeholders' => [
                     {
                     'type'        => 'image',          # Placeholder type image/text
                     'filepath'    => 'test.jpg',       # Image path
                     'width'       => 100,              # Placeholder width
                     'height'      => 300,              # Placeholder height
                     'x'           => 100,              # X cordinates
                     'y'           => 100,              # Y cordinates
                     },
                ]
         }
         ];
  $book->AddLayouts($layouts);
  $book->Create($pdfpath);

=head1 DESCRIPTION

PDF::PhotoBook is an object oriented perl module wich allows you to create custom PDF files with images and text suitable for printing a real photobook.

=head2 METHODS
    
    PDF::PhotoBook::SetTempDir($path)
        Just sets the temporary directory which is used for holding the PDF pades before adding them to the complete output PDF
        These temporary pages are deleted after the job is done. Be careful to set the correct rights to this directory.
        If you did't set it the defaul directory is /tmp but it is recommended to set it manualy

    PDF::PhotoBook::AddLayouts($layouts)
        Just stores the whole configuration into the object

    PDF::PhotoBook::Create($pdfpath)
        Read the whole configuration and create the output pdf

=head2 CONFIGURATION PARAMETERS

    Configuration passed to AddLayouts is an anonymous array with anonymous hashes and should have the following format:
    [
         {'width'       => $pageWidth,  # Page width
         'hieght'       => $pageHeight, # Page height
         'placeholders' => [ # These are all placeholders in the page
                     { # Example image placeholder with full param set
                     'type'        => $type,        # Placeholder type image/text
                     'filepath'    => $imagePath,   # Image path
                     'width'       => $paceWidth,   # Placeholder width
                     'height'      => $placeHeight, # Placeholder height
                     'x'           => $x,           # X cordinates
                     'y'           => $y,           # Y cordinates
                     'mask'        => $mask,        # Optional. This is the path to the mask
                                                    # If we have frame we must apply transparency mask before adding it. 
                     'frame'       => $frame,       # Optional. If we have frame.
                     'imageWidth'  => $imageWidth,  # For resizing the image
                     'imageHeight' => $imageHeight, # For resizing the image 
                     'imageX'      => $imageX,      # Optional. This is used to crop the image if we want to
                     'imageY'      => $imageY,      # Optional. This is used to crop the image if we want to
                     'flop'        => $flop,        # Optional. If we want to flop the image true/false
                     'degrees'     => $degrees,     # Optional. If we need a rotation
                     },
                     { # Example text placeholder with full param set
                     'type'     => 'text',     # Placeholder type image/text
                     'width'    => $width,     # Placeholder width
                     'height'   => $height,    # Placeholder height
                     'x'        => $x,         # X cordinates
                     'y'        => $y,         # Y cordinates
                     'text'     => $text,      # Text content
                     'ttf'      => $fontpath,  # TTF font path
                     'color'    => $fontColor, # Font color
                     'size'     => $fontSize,  # Font size
                     'align'    => $align,     # Optional. Default is left. Text align in the placeholder - left/right/center
                     },
                  ],
         } 
    ]

=head1 SEE ALSO

Image::Magick
PDF::Reuse

http://it-bg.com

=head1 AUTHOR

sickboy, E<lt>sickboy@anarchy.nameE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Ergon

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.


=cut
